esp-hal/src/soc/esp32/psram.rs

```diff
@@ -989,263 +862,251 @@ pub(crate) mod utils {
-            psram_set_qio_read_mode_spi1();
-        }
+    fn psram_cmd_recv_start_spi1(p_rx_data: *mut u32, cmd_mode: PsramCmdMode) {
+        unsafe {
+            let spi = &*crate::peripherals::SPI1::PTR;
+            // get cs1
+            spi.pin().modify(|_, w| w.cs1_dis().clear_bit());
+            spi.pin().modify(|_, w| w.cs0_dis().clear_bit());
+
+            let mode_backup: u32 = (spi.user().read().bits() >> SPI_FWRITE_DUAL_S) & 0xf;
+            let rd_mode_backup: u32 = spi.ctrl().read().bits()
+                & (SPI_FREAD_DIO_M | SPI_FREAD_DUAL_M | SPI_FREAD_QUAD_M | SPI_FREAD_QIO_M);
+
+            if cmd_mode == PsramCmdMode::PsramCmdSpi {
+                psram_set_basic_write_mode_spi1();
+                psram_set_basic_read_mode_spi1();
+            } else if cmd_mode == PsramCmdMode::PsramCmdQpi {
+                psram_set_qio_write_mode_spi1();
+                psram_set_qio_read_mode_spi1();
+            }
 
-        // Wait for SPI0 to idle
-        loop {
-            if read_peri_reg(SPI0_EXT2_REG) == 0 {
-                break;
+            // Wait for SPI0 to idle
+            loop {
+                if spi.ext2().read().bits() == 0 {
+                    break;
+                }
             }
-        }
-        unsafe {
+
             // DPORT_SET_PERI_REG_MASK(DPORT_HOST_INF_SEL_REG, 1 << 14);
             let dport = &*esp32::DPORT::PTR;
             dport
                 .host_inf_sel()
                 .modify(|r, w| w.bits(r.bits() | 1 << 14));
-        }
 
-        // Start send data
-        set_peri_reg_mask(SPI1_CMD_REG, SPI_USR);
-        loop {
-            if read_peri_reg(SPI1_CMD_REG) & SPI_USR == 0 {
-                break;
+            // Start send data
+            spi.cmd().modify(|_, w| w.usr().set_bit());
+            loop {
+                if spi.cmd().read().bits() & SPI_USR == 0 {
+                    break;
+                }
             }
-        }
-        unsafe {
+
             // DPORT_CLEAR_PERI_REG_MASK(DPORT_HOST_INF_SEL_REG, 1 << 14);
             let dport = &*esp32::DPORT::PTR;
             dport
                 .host_inf_sel()
                 .modify(|r, w| w.bits(r.bits() & !(1 << 14)));
-        }
 
-        // recover spi mode
-        set_peri_reg_bits(
-            SPI1_USER_REG,
-            if !p_rx_data.is_null() {
-                SPI_FWRITE_DUAL_M
-            } else {
-                0xf
-            },
-            mode_backup,
-            SPI_FWRITE_DUAL_S,
-        );
-        clear_peri_reg_mask(
-            SPI1_CTRL_REG,
-            SPI_FREAD_DIO_M | SPI_FREAD_DUAL_M | SPI_FREAD_QUAD_M | SPI_FREAD_QIO_M,
-        );
-        set_peri_reg_mask(SPI1_CTRL_REG, rd_mode_backup);
+            // recover spi mode
+            // if !p_rx_data.is_null() {
+            //     spi.user().modify(|_, w| w.fwrite_dual().bits(mode_backup));
+            // } else {
+            //     0xf
+            // }
+            // TODO: get back to this
```

Comment:
Shall we do something about this? Either remove or uncomment (?)

Reply:
Haha, thank you, totally forgot about it! 😆. 
The code in the comment is wrong, I'll remove it. Regarding the function itself (shown below) I don't really understand what's going on here, just as the author of the original commit in IDF is not able to tell me why a write to address 0xf is being made. I would leave the TODO until things are clarified. I don't want to leave a single raw address manipulation in the driver, but holding this PR for an unknown amount of time is not an option either.
```
set_peri_reg_bits(
                SPI1_USER_REG,
                if !p_rx_data.is_null() {
                    SPI_FWRITE_DUAL_M
                } else {
                    0xf
                },
                mode_backup,
                SPI_FWRITE_DUAL_S,
            );
```